File: Data Structures & Algorithms/combinations-of-a-phone-number/submission-0.py
class Solution:
    def letterCombinations(self, digits: str) -> List[str]:
        res=[]
        s=""
        d={"2":"abc","3":"def","4":"ghi","5":"jkl","6":"mno","7":"pqrs","8":"tuv","9":"wxyz"}
        def dfs(i):
            nonlocal s
            if i==len(digits):
                res.append(s[:])
                return
            for _ in d[digits[i]]:
                s+=_
                dfs(i+1)
                s=s[:-1]
        if digits=="":
            return []
        dfs(0)
        return res  

File: Data Structures & Algorithms/partition-labels/submission-0.cpp
class Solution {
public:
    vector<int> partitionLabels(string s) {
        unordered_map<char,int> mp;
        for(int i=0;i<s.size();i++){
            mp[s[i]]=i;
        }    
        int r=mp[s[0]];
        vector<int> res;
        int l=0;
        for(int i=0;i<s.size();i++){
            if(i==r){
                res.push_back(i+1-l);
                l=i+1;
                if(l<s.size())
                    r=mp[s[l]];
            }
            else if(mp[s[i]]>r){
                r=mp[s[i]];
            }
        }
        return res;
    }
};